ddtrace/profiling/collector/_memalloc.c

```diff
@@ -82,41 +79,62 @@ memalloc_add_event(memalloc_context_t* ctx, void* ptr, size_t size)
         uint64_t r = random_range(alloc_count);
         should_sample = r < ctx->max_events;
     }
+    MEMALLOC_GIL_DEBUG_CHECK_RELEASE(&ctx->alloc_gil_guard);
+    return should_sample;
+}
 
-    if (!should_sample) {
-        goto done;
-    }
-
-    traceback_t* tb = memalloc_get_traceback(ctx->max_nframe, ptr, size, ctx->domain);
-    if (!tb) {
-        goto done;
-    }
-
+/* Insert a sample into the profile data structure. The data structure is
+ * protected by the GIL. This function must not call into C Python APIs, which
+ * could release the GIL. Returns a non-NULL traceback if we couldn't add the
+ * sample, either because profiling was stopped or because we are replacing a
+ * sample. The returned traceback should be freed by the caller, since doing so
+ * calls C Python APIs. */
+static traceback_t*
+memalloc_add_sample_no_cpython(memalloc_context_t* ctx, traceback_t* tb)
+{
+    MEMALLOC_GIL_DEBUG_CHECK_ACQUIRE(&ctx->alloc_gil_guard);
     if (!global_alloc_tracker) {
-        /* If getting a traceback lead to a GIL release, it is possible
-         * that the profiler was stopped and the traceback array was freed.
-         * Don't add the sample.
-         */
-        traceback_free(tb);
-        goto done;
+        MEMALLOC_GIL_DEBUG_CHECK_RELEASE(&ctx->alloc_gil_guard);
+        return tb;
     }
 
-    /* The GIL may have been released while collecting a traceback above.
-     * If so, the aggregation/uploading thread may have rotated the tracker.
-     * We need to recompute the index where the sample will go.
-     */
+    traceback_t* old = NULL;
     if (global_alloc_tracker->allocs.count < ctx->max_events) {
         traceback_array_append(&global_alloc_tracker->allocs, tb);
     } else {
         uint64_t r = random_range(ctx->max_events);
-        traceback_t* old = global_alloc_tracker->allocs.tab[r];
-        global_alloc_tracker->allocs.tab[r] = tb;
-        /* Free the old traceback only after modifying the table, in case freeing
+        /* We'll free the old traceback only after modifying the table, in case freeing
```

Comment:
nit: also update the comment to reflect the new structure of the code. After breaking `memalloc_add_event` into few helper functions, including this `memalloc_add_sample_no_cpython`, this function should never call into CPython C API and should never release the GIL. So, the comment here needs to be updated. 

Something along the line of 
'We'll free the old traceback in the caller of this function, `memalloc_add_event`, after modifying the table here. This is because `traceback_free` calls CPython C APIs and they could release the GIL.